File: Ecom/app/src/main/java/com/example/ecom/ItemActivity.java
package com.example.ecom;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.bumptech.glide.Glide;

public class ItemActivity extends AppCompatActivity {

    @SuppressLint("WrongViewCast")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_item);

        ModelClass item = (ModelClass) getIntent().getSerializableExtra("item");

        Log.d("Arya1", "onResponse: " + "Model class working");

        TextView title, price, brand, description, discount, rating;
        ImageView imagesImageView = findViewById(R.id.item_images);

        title = findViewById(R.id.item_title);
        price = findViewById(R.id.item_price);
        brand = findViewById(R.id.item_brand);
        discount = findViewById(R.id.item_discount);
        rating = findViewById(R.id.item_rating);
        description = findViewById(R.id.item_description);
        imagesImageView = findViewById(R.id.item_images);

        if (item != null) {
            title.setText(item.getTitle());
            price.setText(String.valueOf(item.getPrice()));
            brand.setText(item.getBrand());
            discount.setText(String.valueOf(item.getDiscountPercentage()));
            rating.setText(String.valueOf(item.getRating()));
            description.setText(item.getDescription());

            String imagesString = item.getImages();
            imagesString = imagesString.substring(1, imagesString.length() - 1);
            String[] imageUrls = imagesString.replaceAll("\\\\/", "/").replaceAll("\"", "").split(",");

            Log.d("Arya3", "onResponse: " + "Pre condition working");

            for (int i = 0; i < imageUrls.length; i++)
            {
                if (imageUrls.length > 0) {
                    String imageUrl = imageUrls[i];

                    //                images.setText(imageUrl);
                    Glide.with(this)
                            .load(imageUrl)
                            .into(imagesImageView);
                    Log.d("Arya3", "onResponse: " + imageUrl);
                }
            }


        } else {
            Log.d("Arya-sad", "onResponse: " + "NULLLLL");
            title.setText(null);
        }
    }
}

File: QRBUY/app/src/main/java/com/example/qrbuy/ItemActivity.kt
package com.example.qrbuy

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.google.firebase.firestore.FirebaseFirestore

class ItemActivity : AppCompatActivity() {

    private lateinit var itemTitle: TextView
    private lateinit var itemImageView: ImageView
    private lateinit var itemPrice: TextView
    private lateinit var itemaddtoCart: Button
    private lateinit var backButton: ImageButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_item)

        val item = intent.getSerializableExtra("item") as? ModelClass
        val db = FirebaseFirestore.getInstance()

        backButton = findViewById(R.id.backBtn)
        itemTitle = findViewById(R.id.itemTitle)
        itemPrice = findViewById(R.id.itemPrice)
        itemImageView = findViewById(R.id.imageView)
        itemaddtoCart = findViewById(R.id.addtocartbutton)

        if (item != null) {
            itemTitle.setText("Text: "+item.url)
            itemPrice.setText("$"+item.price)
            Glide.with(this)
                .load(item.url)
                .into(itemImageView)

            val DBlocation = db.collection("Users").document(item.userId).collection("cartItems")
            DBlocation
                .whereEqualTo("id", item.id) // Assuming "id" is the field to check for uniqueness
                .get()
                .addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        if (task.result.isEmpty) {
                            itemaddtoCart.isSelected = false
                            itemaddtoCart.text = "Add to Cart"
                        } else {
                            itemaddtoCart.isSelected = true
                            itemaddtoCart.text = "Added to Cart"
                        }
                    }
                }

            itemaddtoCart.setOnClickListener{
                val DBlocation = db.collection("Users").document(item.userId).collection("cartItems")
                DBlocation
                    .whereEqualTo("id", item.id) // Assuming "id" is the field to check for uniqueness
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            if (task.result.isEmpty) {
                                addItemToCart(db, item)
                                itemaddtoCart.isSelected = true
                                itemaddtoCart.text = "Added to Cart"
                            }
                            else {
                                // Item already exists, delete it first
                                val documentSnapshot = task.result.documents[0]
                                val documentId = documentSnapshot.id

                                DBlocation
                                    .document(documentId)
                                    .delete()
                                    .addOnSuccessListener {
                                        itemaddtoCart.isSelected = false
                                        itemaddtoCart.text = "Add to Cart"
                                    }
                                    .addOnFailureListener { e ->
                                        // Handle delete operation failure
                                        Log.e("Firestore", "Error deleting item", e)
                                    }
                            }
                        }
                    }

                Log.d("selected item", item.toString())
            }

            backButton.setOnClickListener{
                finish()
            }

        }
    }

    private fun addItemToCart(db: FirebaseFirestore, model: ModelClass) {
        val UserId: String = model.userId
        val DBlocation = db.collection("Users").document(UserId).collection("cartItems")
        DBlocation
            .add(model)
            .addOnSuccessListener { documentReference ->
                setResult(RESULT_OK)
                Log.d("Firebase", "Item added with ID: ${documentReference.id}")
            }
            .addOnFailureListener { e ->
                Log.e("Firebase", "Error adding item", e)
            }
    }

}
